File: agents/rl_trader.py
import random
from agents.dao_member import DAOMember


class RLTrader(DAOMember):
    """Trader using a simple Q-learning strategy."""

    actions = ("buy", "sell", "add_lp", "remove_lp")

    def __init__(self, unique_id, model, tokens, reputation, location,
                 learning_rate=0.1, discount=0.9, epsilon=0.1, voting_strategy=None):
        super().__init__(unique_id, model, tokens, reputation, location, voting_strategy)
        self.learning_rate = learning_rate
        self.discount = discount
        self.epsilon = epsilon
        self.q = {}
        self.prev_state = None
        self.prev_action = None

    def _state(self):
        price = round(self.model.treasury.get_token_price("DAO_TOKEN"), 2)
        pool = next(iter(self.model.treasury.pools.values()), None)
        depth = 0.0
        if pool:
            depth = pool.reserve_a + pool.reserve_b
        return (price, round(depth, 2))

    def _choose_action(self, state):
        if random.random() < self.epsilon:
            return random.choice(self.actions)
        q_vals = [self.q.get((state, a), 0.0) for a in self.actions]
        return self.actions[q_vals.index(max(q_vals))]

    def _update_q(self, reward, new_state):
        if self.prev_state is None or self.prev_action is None:
            return
        old_q = self.q.get((self.prev_state, self.prev_action), 0.0)
        future = max(self.q.get((new_state, a), 0.0) for a in self.actions)
        self.q[(self.prev_state, self.prev_action)] = old_q + self.learning_rate * (
            reward + self.discount * future - old_q
        )

    def step(self):
        state = self._state()
        action = self._choose_action(state)
        reward = 0.0
        treasury = self.model.treasury
        if action == "buy" and self.tokens > 0:
            amt = min(1, self.tokens)
            treasury.deposit("USDC", amt, step=self.model.current_step)
            out = treasury.swap("USDC", "DAO_TOKEN", amt)
            self.tokens -= amt
            self.tokens += treasury.withdraw("DAO_TOKEN", out, step=self.model.current_step)
            reward = treasury.get_token_price("DAO_TOKEN") * out - amt
        elif action == "sell" and self.tokens > 0:
            amt = min(1, self.tokens)
            treasury.deposit("DAO_TOKEN", amt, step=self.model.current_step)
            out = treasury.swap("DAO_TOKEN", "USDC", amt)
            self.tokens -= amt
            self.tokens += treasury.withdraw("USDC", out, step=self.model.current_step)
            reward = out - treasury.get_token_price("DAO_TOKEN") * amt
        elif action == "add_lp" and self.tokens >= 2:
            treasury.deposit("DAO_TOKEN", 1, step=self.model.current_step)
            treasury.deposit("USDC", 1, step=self.model.current_step)
            treasury.add_liquidity("DAO_TOKEN", "USDC", 1, 1)
            reward = 0.01
        elif action == "remove_lp":
            treasury.remove_liquidity("DAO_TOKEN", "USDC", 0.1)
            reward = -0.01
        self._update_q(reward, state)
        self.prev_state = state
        self.prev_action = action
        self.mark_active()

File: lib/agents/rl-trader.ts
// RL Trader Agent - uses Q-learning to optimize trading strategy
// Refactored to use LearningMixin for standardized Q-learning

import { DAOMember } from './base';
import type { DAOModel } from '../engine/model';
import { logger } from '../utils/logger';
import { LearningMixin, LearningConfig, LearningState } from './learning';
import { StateDiscretizer } from './learning';
import { settings } from '../config/settings';

type Action = 'buy' | 'sell' | 'add_lp' | 'remove_lp' | 'hold';

// Trade sizing
const TRADE_FRACTION = 0.1;  // Trade 10% of holdings
const MIN_TRADE_AMOUNT = 0.1;

export class RLTrader extends DAOMember {
  static readonly ACTIONS: readonly Action[] = ['buy', 'sell', 'add_lp', 'remove_lp', 'hold'];

  // Learning infrastructure
  learning: LearningMixin;

  // Trading state
  prevPrice: number | null = null;
  prevTokens: number;
  tradeCount: number = 0;
  priceHistory: number[] = [];
  private lpTokens: number = 0;

  constructor(
    uniqueId: string,
    model: DAOModel,
    tokens: number = 100,
    reputation: number = 50,
    location: string = 'node_0',
    votingStrategy?: string,
    learningRate?: number,
    discount?: number,
    epsilon?: number
  ) {
    super(uniqueId, model, tokens, reputation, location, votingStrategy);

    // Create learning config from parameters or settings
    const config: Partial<LearningConfig> = {
      learningRate: learningRate ?? settings.learning_global_learning_rate,
      discountFactor: discount ?? settings.learning_discount_factor,
      explorationRate: epsilon ?? settings.learning_exploration_rate,
      explorationDecay: settings.learning_exploration_decay,
      minExploration: settings.learning_min_exploration,
      qBounds: [-50, 50],
    };

    this.learning = new LearningMixin(config);
    this.prevTokens = tokens;
  }

  /**
   * Get current state representation using StateDiscretizer
   * 4-dimensional: price bucket, pool depth, trend, volatility
   */
  private getState(price: number): string {
    if (!this.model.dao) return 'normal|medium|stable|normal';

    // Get pool depth
    const pools = this.model.dao.treasury.pools;
    const firstPool = pools.values().next().value;
    const reserveA = firstPool?.reserveA ?? 0;
    const reserveB = firstPool?.reserveB ?? 0;

    // Use StateDiscretizer for consistent bucketing (3 base dimensions + volatility)
    return StateDiscretizer.combineState(
      StateDiscretizer.discretizePrice(price, 1.0),
      StateDiscretizer.discretizePoolDepth(reserveA, reserveB),
      StateDiscretizer.discretizeTrend(this.priceHistory),
      StateDiscretizer.discretizeVolatility(this.priceHistory)
    );
  }

  /**
   * Execute trading action and return reward
   */
  private executeAction(action: Action): number {
    if (!this.model.dao) return -0.05;

    const treasury = this.model.dao.treasury;
    const primary = this.model.dao.tokenSymbol;
    let reward = 0;

    try {
      switch (action) {
        case 'buy':
          {
            const tradeAmount = Math.max(
              MIN_TRADE_AMOUNT,
              this.getAssetBalance('USDC') * TRADE_FRACTION
            );
          if (this.getAssetBalance('USDC') >= tradeAmount) {
            const debited = this.debitAsset('USDC', tradeAmount);
            if (debited !== tradeAmount) break;
            treasury.deposit('USDC', tradeAmount, this.model.currentStep);
            const out = treasury.swap('USDC', primary, tradeAmount, this.model.currentStep);
            if (out > 0) {
              this.creditAsset(primary, treasury.withdraw(primary, out, this.model.currentStep));
              reward = treasury.getTokenPrice(primary) * out - tradeAmount;
              this.tradeCount++;
            } else {
              this.creditAsset(
                'USDC',
                treasury.withdraw('USDC', tradeAmount, this.model.currentStep)
              );
            }
          }
          break;
          }

        case 'sell':
          {
            const tradeAmount = Math.max(MIN_TRADE_AMOUNT, this.tokens * TRADE_FRACTION);
          if (this.getAssetBalance(primary) >= tradeAmount) {
            const debited = this.debitAsset(primary, tradeAmount);
            if (debited !== tradeAmount) break;
            treasury.deposit(primary, tradeAmount, this.model.currentStep);
            const out = treasury.swap(primary, 'USDC', tradeAmount, this.model.currentStep);
            if (out > 0) {
              this.creditAsset('USDC', treasury.withdraw('USDC', out, this.model.currentStep));
              reward = out - treasury.getTokenPrice(primary) * tradeAmount;
              this.tradeCount++;
            } else {
              this.creditAsset(
                primary,
                treasury.withdraw(primary, tradeAmount, this.model.currentStep)
              );
            }
          }
          break;
          }

        case 'add_lp':
          {
          const lpAmount = Math.min(
            this.getAssetBalance(primary),
            this.getAssetBalance('USDC')
          ) * TRADE_FRACTION;
          if (lpAmount >= MIN_TRADE_AMOUNT) {
            this.debitAsset(primary, lpAmount);
            this.debitAsset('USDC', lpAmount);
            treasury.deposit(primary, lpAmount, this.model.currentStep);
            treasury.deposit('USDC', lpAmount, this.model.currentStep);
            this.lpTokens += treasury.addLiquidity(
              primary,
              'USDC',
              lpAmount,
              lpAmount,
              this.model.currentStep
            );
            reward = 0.02; // Small positive reward for providing liquidity
          }
          break;
          }

        case 'remove_lp': {
          try {
            const lpToBurn = this.lpTokens * 0.1;
            const removed = treasury.removeLiquidityByLP(
              primary,
              'USDC',
              lpToBurn,
              this.model.currentStep
            );
            if (lpToBurn > 0 && (removed[0] > 0 || removed[1] > 0)) {
              const poolKey = [primary, 'USDC'].sort().join('|');
              const pool = treasury.pools.get(poolKey);
              const primaryAmount = pool?.tokenA === primary ? removed[0] : removed[1];
              const usdcAmount = pool?.tokenA === primary ? removed[1] : removed[0];
              this.creditAsset(
                primary,
                treasury.withdraw(primary, primaryAmount, this.model.currentStep)
              );
              this.creditAsset(
                'USDC',
                treasury.withdraw('USDC', usdcAmount, this.model.currentStep)
              );
              this.lpTokens = Math.max(0, this.lpTokens - lpToBurn);
            }
            reward = -0.01; // Small negative reward for removing liquidity
          } catch (error) {
            logger.debug('RLTrader: remove_lp failed', error);
            reward = -0.02; // Penalty for failed action
          }
          break;
        }

        case 'hold': {
          // Calculate reward based on portfolio change
          const portfolioChange = this.tokens - this.prevTokens;
          reward = portfolioChange > 0 ? 0.01 : portfolioChange < 0 ? -0.01 : 0;
          break;
        }
      }
    } catch (error) {
      logger.debug('RLTrader: executeAction failed', error);
      // Action failed, give small negative reward
      reward = -0.05;
    }

    // Clamp reward
    return Math.max(-1, Math.min(1, reward));
  }

  step(): void {
    if (!this.model.dao) return;

    // RL traders participate in governance like other token holders
    this.voteOnRandomProposal();

    // Check if learning is enabled
    if (!settings.learning_enabled) {
      // Fall back to simple hold behavior
      this.markActive();
      return;
    }

    const price = this.model.dao.treasury.getTokenPrice('DAO_TOKEN');

    // Update price history for trend detection
    this.priceHistory.push(price);
    if (this.priceHistory.length > 10) {
      this.priceHistory.shift();
    }

    const state = this.getState(price);

    // Select action using learning mixin
    const action = this.learning.selectAction(
      state,
      [...RLTrader.ACTIONS]
    ) as Action;

    // Execute action and get reward
    const reward = this.executeAction(action);

    // Get next state for Q-update
    const nextPrice = this.model.dao.treasury.getTokenPrice('DAO_TOKEN');
    const nextState = this.getState(nextPrice);

    // Update Q-values
    this.learning.update(
      state,
      action,
      reward,
      nextState,
      [...RLTrader.ACTIONS]
    );

    // Update tracking
    this.prevPrice = price;
    this.prevTokens = this.tokens;
    this.markActive();
  }

  /**
   * Signal end of episode (e.g., simulation reset)
   */
  endEpisode(): void {
    this.learning.endEpisode();
  }

  /**
   * Export learning state for checkpoints
   */
  exportLearningState(): LearningState {
    return this.learning.exportLearningState();
  }

  /**
   * Import learning state from checkpoint
   */
  importLearningState(state: LearningState): void {
    this.learning.importLearningState(state);
  }

  /**
   * Get learning statistics
   */
  getLearningStats(): {
    qTableSize: number;
    stateCount: number;
    episodeCount: number;
    totalReward: number;
    explorationRate: number;
    tradeCount: number;
  } {
    return {
      qTableSize: this.learning.getQTableSize(),
      stateCount: this.learning.getStateCount(),
      episodeCount: this.learning.getEpisodeCount(),
      totalReward: this.learning.getTotalReward(),
      explorationRate: this.learning.getExplorationRate(),
      tradeCount: this.tradeCount,
    };
  }

  // Legacy compatibility getters
  get epsilon(): number {
    return this.learning.getExplorationRate();
  }

  get totalReward(): number {
    return this.learning.getTotalReward();
  }

  get learningRate(): number {
    return this.learning.getConfig().learningRate;
  }

  get discount(): number {
    return this.learning.getConfig().discountFactor;
  }
}
